### ISSUE TITLE
topdirs.cmi multiple locations

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#6754]
**Original bug ID:** 6754
**Reporter:** @ygrek
**Status:** acknowledged (set by @damiendoligez on 2015-02-06T17:43:54Z)
**Resolution:** open
**Priority:** normal
**Severity:** minor
**Version:** 4.02.1
**Target version:** 4.07.0+dev/beta2/rc1/rc2
**Category:** standard library
**Monitored by:** kandu @hhugo @whitequark @gasche @rixed @yallop @hcarty @mmottl

## Bug description

```
$ find /home/ygrek/.opam/4.02.1/lib/ocaml| grep topdirs.cmi
/home/ygrek/.opam/4.02.1/lib/ocaml/compiler-libs/topdirs.cmi
/home/ygrek/.opam/4.02.1/lib/ocaml/topdirs.cmi
```

and every invocation of compiler referencing compiler-libs results in ocamlfind warning
```
findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/ygrek/.opam/4.02.1/lib/ocaml/compiler-libs, /home/ygrek/.opam/4.02.1/lib/ocaml
```

The md5 sum of both modules is the same. I wonder what would be the best way to remove this warning - either the OCaml distribution should install it only in one place or ocamlfind should check md5 and not report it as duplicate because whichever is used will not matter?

**Comment by vicuna** [comment]
**Comment author:** @whitequark

Special-casing the md5 in ocamlfind is of course not the answer. It's not even stable between builds of ocaml.

Why can't topdirs.cmi be installed only in lib/ocaml, anyway?

**Comment by vicuna** [comment]
**Comment author:** @ygrek

I mean that once ocamlfind detects same cmi occuring in multiple dirs it could check the md5 of those cmi and if they all are equal - it means interfaces are really equal.
But I agree that installing in only one location would be the best solution.

**Comment by vicuna** [comment]
**Comment author:** @damiendoligez

Interesting. My first reaction was put it into lib/ocaml/compiler-libs, where you also find toploop, topmain, and topstart. @whitequark, do you have any particular reason to want it in lib/ocaml ?

**Comment by vicuna** [comment]
**Comment author:** @whitequark

I thought topfind uses Topdirs but I now see that it does #directory "+compiler-libs" so that doesn't matter anyway.

**Comment by vicuna** [comment]
**Comment author:** @hcarty

Is it safe to move/remove lib/ocaml/topdirs.cmi?  Anything using ppx_blob, for example, generates lots of findlib warnings.

Moving that file and adding '#directory "+compiler-libs";;' to my ~/.ocamlinit allowed utop to work and code to compile without the findlib warnings.

**Comment by vicuna** [comment]
**Comment author:** @whitequark

@doligez, right, hcarty's comment above is why I wanted it in lib/ocaml, I think

**Comment by vicuna** [comment]
**Comment author:** @hhugo

Any chance to see this resolved for 4.03 ?

**Comment by vicuna** [comment]
**Comment author:** @johnwhitington

This warning is, by default, fatal to ocamlmktop, so one has to do "-warn-error -31" if making a custom top level with compiler-libs.

Does anyone know why warning 31 should be fatal for ocamlmktop but not ocamlc?

**Comment by vicuna** [comment]
**Comment author:** pveber

Removing lib/ocaml/topdirs.cmi seems the cleanest solution to me. For the record, this would require some adjustments in opam (for the generated .ocamlinit) and oasis (the generated setup.ml file in dynamic update mode) and maybe other libs. Before doing anything, it would wiser I think to announce/discuss it on the list, so that everybody can think and report on the possible implications.

In any case, it would be great to fix this. I can provide some help if needed.

**Comment by vicuna** [comment]
**Comment author:** @gasche

We missed the timeline for 4.03 on this one. If someone (Philippe included) is willing to take action on this issue (starting a mailing-list discussion etc.), now would be an excellent time, so that we get a change early in the 4.04+dev lifetime.

**Comment by olafhering** [comment]
Likely related [bug#1156934 ocaml-runtime requires ocaml-compiler-libs-devel](https://bugzilla.suse.com/show_bug.cgi?id=1156934)

**Comment by nojb** [comment]
Anyone knows what is the status of this? It appears it is still the case that `topdirs.cmi` is installed twice in the installation directory and I was wondering if there was a specific reason for this?

**Comment by gasche** [comment]
I think that the two copies were left around for compatibility (originally it was just in `ocaml`, and when we introduced `ocaml/compiler-libs` it was also installed there). As the discussion above shows, it's not completely obvious to decide which of the two places should be favored. I may be fine  to remove today the `ocaml/topdirs.cmi` copy; it may break stuff, but the `+compiler-libs` fix is backward-compatible with all recent-enough versions of OCaml (which was not the case originally).

**Comment by nojb** [comment]
> I may be fine to remove today the `ocaml/topdirs.cmi` copy; it may break stuff, but the `+compiler-libs` fix is backward-compatible with all recent-enough versions of OCaml (which was not the case originally).

Thanks; an even better solutions would be provided by #9056 as we can make just the one file visible to the compiler, see #5628.

**Comment by nojb** [comment]
I think this was fixed by #11199.
